datafusion/optimizer/src/single_distinct_to_groupby.rs

```diff
@@ -131,177 +126,190 @@ fn contains_grouping_set(expr: &[Expr]) -> bool {
+                            (
+                                group_expr.alias(alias_str.clone()),
                                 (
-                                    alias_expr,
-                                    (
-                                        col(alias_str),
-                                        Some(qualified_name(qualifier, field.name())),
-                                    ),
-                                )
-                            }
-                        })
-                        .unzip();
-
-                    // and they can be referenced by the alias in the outer aggr plan
-                    let outer_group_exprs = out_group_expr_with_alias
-                        .iter()
-                        .map(|(out_group_expr, _)| out_group_expr.clone())
-                        .collect::<Vec<_>>();
-
-                    // replace the distinct arg with alias
-                    let mut index = 1;
-                    let mut group_fields_set = HashSet::new();
-                    let mut inner_aggr_exprs = vec![];
-                    let outer_aggr_exprs = aggr_expr
-                        .iter()
-                        .map(|aggr_expr| match aggr_expr {
-                            Expr::AggregateFunction(AggregateFunction {
-                                func_def: AggregateFunctionDefinition::BuiltIn(fun),
-                                args,
-                                distinct,
-                                ..
-                            }) => {
-                                // is_single_distinct_agg ensure args.len=1
-                                if *distinct
-                                    && group_fields_set.insert(args[0].display_name()?)
-                                {
-                                    inner_group_exprs.push(
-                                        args[0].clone().alias(SINGLE_DISTINCT_ALIAS),
-                                    );
+                                    col(alias_str),
+                                    Some(qualified_name(qualifier, field.name())),
+                                ),
+                            )
+                        }
+                    })
+                    .unzip();
+
+                // replace the distinct arg with alias
+                let mut index = 1;
+                let mut distinct_aggr_exprs = HashSet::new();
+                let mut inner_aggr_exprs = vec![];
+                let outer_aggr_exprs = aggr_expr
+                    .into_iter()
+                    .map(|aggr_expr| match aggr_expr {
+                        Expr::AggregateFunction(AggregateFunction {
+                            func_def: AggregateFunctionDefinition::BuiltIn(fun),
+                            mut args,
+                            distinct,
+                            ..
+                        }) => {
+                            if distinct {
+                                debug_assert_eq!(
```

Comment:
Can we please change this from `debug_assert_eq` (which will panic) to an `internal_err!`?

Reply:
Yes, return an `internal_err` gives users some idea why failing. Updated in https://github.com/apache/datafusion/pull/10527/commits/55e3d16fbd89b17574d19041c6f752268915a139